hgapp/crafting/views.py

```diff
@@ -268,9 +273,47 @@ def __get_page_data_and_forms(self, POST=None, early_access=False):
         all_contractor_power_full_ids = set(power_fulls.values_list('id', flat=True))
         additional_artifact_power_full_ids = set()
         for artifact in crafted_artifacts:
+
             current_fulls = set(artifact.power_full_set.values_list('id', flat=True))
             additional_artifact_power_full_ids.update(current_fulls.difference(all_contractor_power_full_ids))
             refundable_fulls = refundable_power_fulls_by_artifact_id[artifact.pk]
+
+            # Get upgradable ones... so, all of powers
+            upgradable_pks = all_contractor_power_full_ids
+            # That are on this artifact
+            upgradable_pks = upgradable_pks.intersection(current_fulls)
+            # That aren't refundable... if they're refundable they'll be auto-refunded if they enter a condition where
+            # they could be otherwise upgraded.
+            # Correction, I need to know if it's being updated in higher level logic.
+            # upgradable_pks = upgradable_pks.difference(refundable_fulls)
+
+            power_revisions_on_artifact_by_parent_id = defaultdict(list)
+            for power_revision in artifact.power_set.all():
+                # We only want the latest, but this should be in order...
+                power_revisions_on_artifact_by_parent_id[power_revision.parent_power_id] = power_revision
+
+                # For each power revision on the artifact, if its revision is equal to the latest revision's revision
+                # It's not ugpradable.
+                if power_revision.parent_power_id in all_crafter_powers_full_object_by_artifact_id:
+                    if power_revision.id == all_crafter_powers_full_object_by_artifact_id[power_revision.parent_power.id].latest_rev_id:
+                        if power_revision.parent_power_id in upgradable_pks:
+                            # Need to be able to
+                            if power_revision.parent_power_id not in refundable_fulls:
+                                upgradable_pks.remove(power_revision.parent_power_id)
+
+            # And the cost.
+            # Going to do a map of upgradable power -> cost.  This overrides the values in the crafting blob pk list.
+            upgradable_costs_by_pk = defaultdict()
+            for upgradablePower in upgradable_pks:
+                # This is true when we're refunding
+                if all_crafter_powers_full_object_by_artifact_id[upgradablePower].latest_rev == power_revisions_on_artifact_by_parent_id[upgradablePower]:
+                    previous_revisions = artifact.power_set.filter(parent_power_id=upgradablePower).order_by('-pub_date')
+                    if previous_revisions.count() >= 2:
+                        previous_revision_for_upgradable_power = previous_revisions[1]
+                        upgradable_costs_by_pk[upgradablePower] = all_crafter_powers_full_object_by_artifact_id[upgradablePower].get_gift_cost_delta(previous_revision_for_upgradable_power) + 1
```

Comment:
Since you reference this +1 multiple times (including in `get_exp_cost_for_crafting_instance`) it might as well be a constant in models.py